### ISSUE TITLE
GAIL performs worse than Behavioral Cloning

### ISSUE THREAD

**Issue by engyasin** [HumanCompatibleAI/imitation#582]
## Bug description
I was testing the implementation of GAIL on some simple environments with big action space : Discrete(20) , and the results were not always much better (or even worse) than BC.
Is that normal? or maybe I just had the wrong hyperparameters? 
I repeated the tests several times but the improvement is very slow and very littlie.
I'm surprised just because it's known that GAIL has better results than Behavior cloning. 

I didn't test too much on other envs, but according to your documentation, the situation is the same, on 'CartPole-v1'.

**Comment by AdamGleave** [comment]
BC can be a strong baseline in simple environments when there are plentiful demonstrations. Indeed I've seen many GAIL (and other imitation learning) implementations warm-start the policy from the BC policy -- which goes some way to ensuring they don't do worse than BC.

@taufeeque9 is currently working on hyperparameter tuning and benchmarking in a range of environments, he can maybe share some preliminary results, but we'll have a better sense of this once that exercise is complete.

If you want to work on this then more eyes are definitely welcome. For tuning GAIL I'd suggest starting by getting hyperparameters for the RL algorithm that work really well on a ground-truth reward (if you don't have one, then tune on a similar environment that does). Then could do a sweep to see performance as you vary # of demos of both BC and GAIL. It'd also be helpful to evaluate other implementations of GAIL, like SB2's GAIL, on the same environments.

Are you environments fixed horizon? The original GAIL paper's results were somewhat confounded by the environments being variable horizon and there being a hyperparameter that chose the sign of the reward as appropriate for that environment (whether you want early termination or not), which [Kostrikov et al](https://arxiv.org/abs/1809.02925) showed leads to good performance even in the absence of any demonstrations! So unfortunately some of the published GAIL results may just not replicate.

**Comment by engyasin** [comment]
Thanks @AdamGleave  for the explanation, 

To give more details, I was testing on Particle Environment https://pettingzoo.farama.org/environments/mpe/simple_spread/ , but instead of comparing the reward, (where GAIL and BC have similar results) , I tested the actions matching. So the trained policy and the expert policy are tested to see if their actions match in response to the same state input, which is the whole point of imitation learning , i.e to copy the policy not just high rewards (it might be equivalent if the expert is optimal, but this is additional assumption) 

The results were double the number for BC (50%) , while GAIL only 28%.
The only explanation I can think of now is that the environment is so simple that BC can master from the demos.

I think the environment is with variable horizon, but I'm not sure how that affect GAIL, so I have to check that. As well as testing on other envs/ implementation of GAIL

**Comment by AdamGleave** [comment]
Comparing similarity of action distributions is a reasonable metric but it's not one I've seen GAIL evaluated on before, so very much an open question how well it does on that. BC is directly optimizing for choosing the same actions so might have an edge on that metric.

You might want to also look at similarity of the state distribution. With BC if you get the same action as the expert 99% of the time you can still drift further and further away from where the expert would be each time you take the 1% "wrong" actions and get low reward as a result. Whereas methods like GAIL are more trying to mimic the state distribution, so if they take an action different to what the expert would the policy learns to "steer" back to where the expert would have gone, mitigating the DAgger problem.

**Comment by engyasin** [comment]
Yeah, I think you're right. BC have problems with drifting.

Feel free to close the issue if required, because it maybe not related to the code.  But I can also put my future test results here in case someone else has the same question, even if it's closed.

**Comment by AdamGleave** [comment]
Yeah I'm going to close but feel free to update the issue, I'm curious to see how you get on!

**Comment by engyasin** [comment]
After testing on the special case of 1-agent, GAIL was able to outperform BC in tens of thousands of steps.

After checking the literature on the required number of steps for MPE environment, usually it is 10M.

As a result BC train fast till good accuracy, however if more accuracy is wanted then training GAIL will take much time but it will give better results.

**Comment by AdamGleave** [comment]
Thanks for the update! That makes sense to me, GAIL will definitely be slower to train as it runs an RL algorithm in the inner loop.
